common/models/user.js

```diff
@@ -336,8 +330,24 @@ module.exports = function(User) {
   User.prototype.hasPassword = function(plain, fn) {
     fn = fn || utils.createPromiseCallback();
     if (this.password && plain) {
-      bcrypt.compare(plain, this.password, function(err, isMatch) {
+      var hashed = this.password;
+      var lengthWasValidated = false;
+      if (/^\$72\$/.test(hashed) && hashed.length === 60) {
+        hashed = '$2a$'.concat(hashed.slice(4));
+        lengthWasValidated = true;
+      }
+      bcrypt.compare(plain, hashed, function(err, isMatch) {
         if (err) return fn(err);
+
+        // Only check password length after we run the same code we would execute
+        // for a password with a valid length. This should prevent attackers from guessing
+        // the maximum allowed password length by measuring the response time.
+        if (plain.length > MAX_PASSWORD_LENGTH && lengthWasValidated) {
+          // Reject too long password only if we are sure that the real
+          // password is at most 72 chars long.
+          fn(null, false);
```

Comment:
Ha, this will call `fn` twice, should be `return fn(null, false)`.

@loay I'll leave it up to you to fix this.